Missing Package Source Mapping in NuGet.config

### Description

When developing on macOS, packageSourceMapping is required as this feature is enabled per default. Until that is set up in the NuGet.config it is not possible to restore NuGet pckages.

### Steps to Reproduce

1. Clone latest main
2. `dotnet restore Microsoft.Maui-mac.slnf`

Doing the above gives you a huge red wall of text with errors failing to restore packages.

Log: https://gist.github.com/Cheesebaron/dd1977ee587e62b3241db3abc380f783



### Link to public reproduction project repository

https://github.com/dotnet/maui

### Version with bug

7.0.49

### Last version that worked well

Unknown/Other

### Affected platforms

macOS

### Affected platform versions

macOS

### Did you find any workaround?

_No response_

### Relevant log output

_No response_

@drasticactions even running the commands in the DEVELOPMENT doc fail:

Running `dotnet tool restore`:

```
/var/folders/s9/3m9n8cmx5gqdfd60777_3wbm0000gn/T/955c2402-ea3f-40fb-b18b-4ee02e96f8b1/restore.csproj : error NU1100: Unable to resolve 'cake.tool (= 2.3.0)' for 'net7.0'. PackageSourceMapping is enabled, the following source(s) were not considered: /usr/local/share/dotnet/library-packs, benchmark-dotnet-prerelease, darc-pub-dotnet-emsdk-c3fc739, darc-pub-dotnet-runtime-1e1f688, dotnet-eng, dotnet-public, dotnet6, dotnet7, skiasharp.
/var/folders/s9/3m9n8cmx5gqdfd60777_3wbm0000gn/T/955c2402-ea3f-40fb-b18b-4ee02e96f8b1/restore.csproj : error NU1100: Unable to resolve 'cake.tool (= 2.3.0)' for 'net7.0/any'. PackageSourceMapping is enabled, the following source(s) were not considered: /usr/local/share/dotnet/library-packs, benchmark-dotnet-prerelease, darc-pub-dotnet-emsdk-c3fc739, darc-pub-dotnet-runtime-1e1f688, dotnet-eng, dotnet-public, dotnet6, dotnet7, skiasharp.
/var/folders/s9/3m9n8cmx5gqdfd60777_3wbm0000gn/T/7b6479da-78f6-4f9c-97da-c87ef9f13573/restore.csproj : error NU1100: Unable to resolve 'powershell (= 7.3.6)' for 'net7.0'. PackageSourceMapping is enabled, the following source(s) were not considered: /usr/local/share/dotnet/library-packs, benchmark-dotnet-prerelease, darc-pub-dotnet-emsdk-c3fc739, darc-pub-dotnet-runtime-1e1f688, dotnet-eng, dotnet-public, dotnet6, dotnet7, skiasharp.
/var/folders/s9/3m9n8cmx5gqdfd60777_3wbm0000gn/T/7b6479da-78f6-4f9c-97da-c87ef9f13573/restore.csproj : error NU1100: Unable to resolve 'powershell (= 7.3.6)' for 'net7.0/any'. PackageSourceMapping is enabled, the following source(s) were not considered: /usr/local/share/dotnet/library-packs, benchmark-dotnet-prerelease, darc-pub-dotnet-emsdk-c3fc739, darc-pub-dotnet-runtime-1e1f688, dotnet-eng, dotnet-public, dotnet6, dotnet7, skiasharp.
/var/folders/s9/3m9n8cmx5gqdfd60777_3wbm0000gn/T/eea1c29b-f8c8-4001-aa3e-260b64942aab/restore.csproj : error NU1100: Unable to resolve 'microsoft.dotnet.xharness.cli (= 8.0.0-prerelease.23407.2)' for 'net7.0'. PackageSourceMapping is enabled, the following source(s) were not considered: /usr/local/share/dotnet/library-packs, benchmark-dotnet-prerelease, darc-pub-dotnet-emsdk-c3fc739, darc-pub-dotnet-runtime-1e1f688, dotnet-eng, dotnet-public, dotnet6, dotnet7, skiasharp.
/var/folders/s9/3m9n8cmx5gqdfd60777_3wbm0000gn/T/eea1c29b-f8c8-4001-aa3e-260b64942aab/restore.csproj : error NU1100: Unable to resolve 'microsoft.dotnet.xharness.cli (= 8.0.0-prerelease.23407.2)' for 'net7.0/any'. PackageSourceMapping is enabled, the following source(s) were not considered: /usr/local/share/dotnet/library-packs, benchmark-dotnet-prerelease, darc-pub-dotnet-emsdk-c3fc739, darc-pub-dotnet-runtime-1e1f688, dotnet-eng, dotnet-public, dotnet6, dotnet7, skiasharp.

Package "cake.tool" failed to restore, due to Microsoft.DotNet.ToolPackage.ToolPackageException: The tool package could not be restored.
   at Microsoft.DotNet.Tools.Tool.Install.ProjectRestorer.Restore(FilePath project, PackageLocation packageLocation, String verbosity)
   at Microsoft.DotNet.ToolPackage.ToolPackageInstaller.InstallPackageToExternalManagedLocation(PackageLocation packageLocation, PackageId packageId, VersionRange versionRange, String targetFramework, String verbosity)
   at Microsoft.DotNet.Tools.Tool.Restore.ToolRestoreCommand.InstallPackages(ToolManifestPackage package, Nullable`1 configFile)
Package "powershell" failed to restore, due to Microsoft.DotNet.ToolPackage.ToolPackageException: The tool package could not be restored.
   at Microsoft.DotNet.Tools.Tool.Install.ProjectRestorer.Restore(FilePath project, PackageLocation packageLocation, String verbosity)
   at Microsoft.DotNet.ToolPackage.ToolPackageInstaller.InstallPackageToExternalManagedLocation(PackageLocation packageLocation, PackageId packageId, VersionRange versionRange, String targetFramework, String verbosity)
   at Microsoft.DotNet.Tools.Tool.Restore.ToolRestoreCommand.InstallPackages(ToolManifestPackage package, Nullable`1 configFile)
Package "microsoft.dotnet.xharness.cli" failed to restore, due to Microsoft.DotNet.ToolPackage.ToolPackageException: The tool package could not be restored.
   at Microsoft.DotNet.Tools.Tool.Install.ProjectRestorer.Restore(FilePath project, PackageLocation packageLocation, String verbosity)
   at Microsoft.DotNet.ToolPackage.ToolPackageInstaller.InstallPackageToExternalManagedLocation(PackageLocation packageLocation, PackageId packageId, VersionRange versionRange, String targetFramework, String verbosity)
   at Microsoft.DotNet.Tools.Tool.Restore.ToolRestoreCommand.InstallPackages(ToolManifestPackage package, Nullable`1 configFile)

Tool 'api-tools' (version '1.3.5') was restored. Available commands: api-tools
```

So yes...

--- a/NuGet.config
+++ b/NuGet.config
@@ -26,6 +26,7 @@
     </packageSource> -->
     <packageSource key="dotnet-eng">
       <package pattern="microsoft.dotnet.*" />
+      <package pattern="Microsoft.DiaSymReader.*" />
     </packageSource>
     <packageSource key="dotnet-public">
       <package pattern="*" />

--- a/NuGet.config
+++ b/NuGet.config
@@ -1,5 +1,8 @@
 <?xml version="1.0" encoding="utf-8"?>
 <configuration>
+
+  <!-- Define the package sources: all mirrors or internal feeds. -->
+  <!-- `clear` ensures no additional sources are inherited from another config file. -->
   <packageSources>
     <clear />
     <!-- <add key="local" value="artifacts" /> -->
@@ -8,8 +11,27 @@
     <add key="dotnet-public" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet-public/nuget/v3/index.json" protocolVersion="3" />
     <add key="benchmark-dotnet-prerelease" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/benchmark-dotnet-prerelease/nuget/v3/index.json" />
   </packageSources>
+
   <activePackageSource>
     <add key="All" value="(Aggregate source)" />
   </activePackageSource>
+
   <disabledPackageSources />
+
+  <!-- Define mappings by adding package patterns beneath the target source. -->
+  <!-- Some packages will restore from different locations, but most will be from `dotnet-public`. -->
+  <packageSourceMapping>
+    <!-- key value for <packageSource> should match key values from <packageSources> element -->
+    <!-- <packageSource key="dotnet-tools">
+    </packageSource> -->
+    <packageSource key="dotnet-eng">
+      <package pattern="microsoft.dotnet.xharness.*" />
+    </packageSource>
+    <packageSource key="dotnet-public">
+      <package pattern="*" />
+    </packageSource>
+    <!-- <packageSource key="benchmark-dotnet-prerelease">
+    </packageSource> -->
+  </packageSourceMapping>
+
 </configuration>

--- a/NuGet.config
+++ b/NuGet.config
@@ -5,7 +5,8 @@
   <!-- `clear` ensures no additional sources are inherited from another config file. -->
   <packageSources>
     <clear />
-    <!-- <add key="local" value="artifacts" /> -->
+    <add key="local" value="LOCAL_PLACEHOLDER" />
+    <add key="nuget-only" value="NUGET_ONLY_PLACEHOLDER" />
     <add key="dotnet-tools" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet-tools/nuget/v3/index.json" protocolVersion="3" />
     <add key="dotnet-eng" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet-eng/nuget/v3/index.json" protocolVersion="3" />
     <add key="dotnet-public" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet-public/nuget/v3/index.json" protocolVersion="3" />
@@ -17,12 +18,23 @@
     <add key="All" value="(Aggregate source)" />
   </activePackageSource>
 
-  <disabledPackageSources />
+  <disabledPackageSources>
+    <add key="local" value="true" />
+    <add key="nuget-only" value="true" />
+  </disabledPackageSources>
 
   <!-- Define mappings by adding package patterns beneath the target source. -->
   <!-- Some packages will restore from different locations, but most will be from `dotnet-public`. -->
+  <!-- The key value for <packageSource> should match key values from <packageSources> element. -->
   <packageSourceMapping> 
-    <!-- key value for <packageSource> should match key values from <packageSources> element -->
+    <packageSource key="local">
+      <package pattern="Microsoft.Maui.*" />
+      <package pattern="Microsoft.AspNetCore.Components.WebView.*" />
+    </packageSource>
+    <packageSource key="nuget-only">
+      <package pattern="Microsoft.Maui.*" />
+      <package pattern="Microsoft.AspNetCore.Components.WebView.*" />
+    </packageSource>
     <!-- <packageSource key="dotnet-tools">
     </packageSource> -->
     <packageSource key="dotnet-eng">

--- a/NuGet.config
+++ b/NuGet.config
@@ -31,6 +31,7 @@
     </packageSource>
     <packageSource key="dotnet-public">
       <package pattern="*" />
+      <package pattern="Microsoft.DotNet.PlatformAbstractions" />
     </packageSource>
     <!-- <packageSource key="benchmark-dotnet-prerelease">
     </packageSource> -->

--- a/NuGet.config
+++ b/NuGet.config
@@ -44,6 +44,8 @@
     <packageSource key="dotnet-public">
       <package pattern="*" />
       <package pattern="Microsoft.DotNet.PlatformAbstractions" />
+      <package pattern="Microsoft.Maui.*" />
+      <package pattern="Microsoft.AspNetCore.Components.WebView.*" />
     </packageSource>
     <!-- <packageSource key="benchmark-dotnet-prerelease">
     </packageSource> -->

--- a/NuGet.config
+++ b/NuGet.config
@@ -10,6 +10,7 @@
     <add key="dotnet-eng" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet-eng/nuget/v3/index.json" protocolVersion="3" />
     <add key="dotnet-public" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet-public/nuget/v3/index.json" protocolVersion="3" />
     <add key="benchmark-dotnet-prerelease" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/benchmark-dotnet-prerelease/nuget/v3/index.json" />
+    <add key="skiasharp" value="https://pkgs.dev.azure.com/xamarin/public/_packaging/SkiaSharp/nuget/v3/index.json" />
   </packageSources>
 
   <activePackageSource>
@@ -20,7 +21,7 @@
 
   <!-- Define mappings by adding package patterns beneath the target source. -->
   <!-- Some packages will restore from different locations, but most will be from `dotnet-public`. -->
-  <packageSourceMapping>
+  <packageSourceMapping> 
     <!-- key value for <packageSource> should match key values from <packageSources> element -->
     <!-- <packageSource key="dotnet-tools">
     </packageSource> -->
@@ -33,6 +34,9 @@
     </packageSource>
     <!-- <packageSource key="benchmark-dotnet-prerelease">
     </packageSource> -->
+    <packageSource key="skiasharp">
+      <package pattern="_NativeAssets.*" />
+    </packageSource>
   </packageSourceMapping>
 
 </configuration>

--- a/NuGet.config
+++ b/NuGet.config
@@ -22,11 +22,10 @@
   <!-- Some packages will restore from different locations, but most will be from `dotnet-public`. -->
   <packageSourceMapping>
     <!-- key value for <packageSource> should match key values from <packageSources> element -->
-    <packageSource key="dotnet-tools">
-      <package pattern="Microsoft.DotNet.Arcade.*" />
-    </packageSource>
+    <!-- <packageSource key="dotnet-tools">
+    </packageSource> -->
     <packageSource key="dotnet-eng">
-      <package pattern="microsoft.dotnet.xharness.*" />
+      <package pattern="microsoft.dotnet.*" />
     </packageSource>
     <packageSource key="dotnet-public">
       <package pattern="*" />

--- a/NuGet.config
+++ b/NuGet.config
@@ -22,8 +22,9 @@
   <!-- Some packages will restore from different locations, but most will be from `dotnet-public`. -->
   <packageSourceMapping>
     <!-- key value for <packageSource> should match key values from <packageSources> element -->
-    <!-- <packageSource key="dotnet-tools">
-    </packageSource> -->
+    <packageSource key="dotnet-tools">
+      <package pattern="Microsoft.DotNet.Arcade.*" />
+    </packageSource>
     <packageSource key="dotnet-eng">
       <package pattern="microsoft.dotnet.xharness.*" />
     </packageSource>

--- a/NuGet.config
+++ b/NuGet.config
@@ -46,6 +46,7 @@
       <package pattern="Microsoft.DotNet.PlatformAbstractions" />
       <package pattern="Microsoft.Maui.*" />
       <package pattern="Microsoft.AspNetCore.Components.WebView.*" />
+      <package pattern="Microsoft.DotNet.ILCompiler" />
     </packageSource>
     <!-- <packageSource key="benchmark-dotnet-prerelease">
     </packageSource> -->

--- a/eng/cake/dotnet.cake
+++ b/eng/cake/dotnet.cake
@@ -28,6 +28,7 @@ var NuGetOnlyPackages = new string[] {
     "Microsoft.Maui.Essentials.*.{nupkg,snupkg}",
     "Microsoft.Maui.Graphics.*.{nupkg,snupkg}",
     "Microsoft.Maui.Maps.*.{nupkg,snupkg}",
+    "Microsoft.Maui.Resizetizer.{nupkg,snupkg}",
     "Microsoft.AspNetCore.Components.WebView.*.{nupkg,snupkg}",
 };
 

--- a/eng/cake/dotnet.cake
+++ b/eng/cake/dotnet.cake
@@ -304,7 +304,7 @@ Task("dotnet-pack-additional")
             Version = nativeAssetsVersion,
             ExcludeVersion = true,
             OutputDirectory = assetsDir,
-            Source = new[] { "https://aka.ms/skiasharp-eap/index.json" },
+            Source = new[] { "https://pkgs.dev.azure.com/xamarin/public/_packaging/SkiaSharp/nuget/v3/index.json" },
         });
         foreach (var nupkg in GetFiles($"{assetsDir}/**/*.nupkg"))
             DeleteFile(nupkg);

--- a/eng/cake/dotnet.cake
+++ b/eng/cake/dotnet.cake
@@ -28,7 +28,7 @@ var NuGetOnlyPackages = new string[] {
     "Microsoft.Maui.Essentials.*.{nupkg,snupkg}",
     "Microsoft.Maui.Graphics.*.{nupkg,snupkg}",
     "Microsoft.Maui.Maps.*.{nupkg,snupkg}",
-    "Microsoft.Maui.Resizetizer.{nupkg,snupkg}",
+    "Microsoft.Maui.Resizetizer.*.{nupkg,snupkg}",
     "Microsoft.AspNetCore.Components.WebView.*.{nupkg,snupkg}",
 };
 

--- a/eng/cake/dotnet.cake
+++ b/eng/cake/dotnet.cake
@@ -47,10 +47,8 @@ Task("dotnet")
         {
             EnsureDirectoryExists(nugetSource);
             var originalNuget = File("./NuGet.config");
-            ReplaceTextInFiles(
-                originalNuget,
-                 $"<!-- <add key=\"local\" value=\"artifacts\" /> -->",
-                $"<add key=\"nuget-only\" value=\"{nugetSource}\" />");
+            ReplaceTextInFiles(originalNuget, "<add key=\"nuget-only\" value=\"true\" />", "");
+            ReplaceTextInFiles(originalNuget, "NUGET_ONLY_PLACEHOLDER", nugetSource);
         }
 
         DotNetBuild("./src/DotNet/DotNet.csproj", new DotNetBuildSettings
@@ -267,10 +265,8 @@ Task("dotnet-pack-maui")
         {
             EnsureDirectoryExists(nugetSource);
             var originalNuget = File("./NuGet.config");
-            ReplaceTextInFiles(
-                originalNuget,
-                $"<!-- <add key=\"local\" value=\"artifacts\" /> -->",
-                $"<add key=\"local\" value=\"{nugetSource}\" />");
+            ReplaceTextInFiles(originalNuget, "<add key=\"local\" value=\"true\" />", "");
+            ReplaceTextInFiles(originalNuget, "LOCAL_PLACEHOLDER", nugetSource);
         }
 
         var sln = "./Microsoft.Maui.Packages.slnf";
@@ -800,10 +796,8 @@ DirectoryPath PrepareSeparateBuildContext(string dirName, string props = null, s
     }
 
     // Add a specific folder for nuget-only packages
-    ReplaceTextInFiles(
-        config.FullPath,
-        $"<!-- <add key=\"local\" value=\"artifacts\" /> -->",
-        $"<add key=\"nuget-only\" value=\"{nugetOnly.FullPath}\" />");
+    ReplaceTextInFiles(config.FullPath, "<add key=\"nuget-only\" value=\"true\" />", "");
+    ReplaceTextInFiles(config.FullPath, "NUGET_ONLY_PLACEHOLDER", nugetOnly.FullPath);
 
     // Create empty or copy test Directory.Build.props/targets
     if (string.IsNullOrEmpty(props))

--- a/src/TestUtils/src/Microsoft.Maui.IntegrationTests/BaseBuildTest.cs
+++ b/src/TestUtils/src/Microsoft.Maui.IntegrationTests/BaseBuildTest.cs
@@ -67,7 +67,7 @@ namespace Microsoft.Maui.IntegrationTests
 				"Microsoft.Maui.Essentials.*.nupkg",
 				"Microsoft.Maui.Graphics.*.nupkg",
 				"Microsoft.Maui.Maps.*.nupkg",
-				"Microsoft.Maui.Resizetizer.nupkg",
+				"Microsoft.Maui.Resizetizer.*.nupkg",
 				"Microsoft.AspNetCore.Components.WebView.*.nupkg",
 			};
 

--- a/src/TestUtils/src/Microsoft.Maui.IntegrationTests/BaseBuildTest.cs
+++ b/src/TestUtils/src/Microsoft.Maui.IntegrationTests/BaseBuildTest.cs
@@ -87,9 +87,8 @@ namespace Microsoft.Maui.IntegrationTests
 			}
 
 			File.Copy(Path.Combine(TestEnvironment.GetMauiDirectory(), "NuGet.config"), TestNuGetConfig, true);
-			FileUtilities.ReplaceInFile(TestNuGetConfig,
-				"<!-- <add key=\"local\" value=\"artifacts\" /> -->",
-				$"<add key=\"nuget-only\" value=\"{extraPacksDir}\" />");
+			FileUtilities.ReplaceInFile(TestNuGetConfig, "<add key=\"nuget-only\" value=\"true\" />", "");
+			FileUtilities.ReplaceInFile(TestNuGetConfig, "NUGET_ONLY_PLACEHOLDER", extraPacksDir);
 		}
 
 		[SetUp]

--- a/src/TestUtils/src/Microsoft.Maui.IntegrationTests/BaseBuildTest.cs
+++ b/src/TestUtils/src/Microsoft.Maui.IntegrationTests/BaseBuildTest.cs
@@ -67,6 +67,7 @@ namespace Microsoft.Maui.IntegrationTests
 				"Microsoft.Maui.Essentials.*.nupkg",
 				"Microsoft.Maui.Graphics.*.nupkg",
 				"Microsoft.Maui.Maps.*.nupkg",
+				"Microsoft.Maui.Resizetizer.nupkg",
 				"Microsoft.AspNetCore.Components.WebView.*.nupkg",
 			};
 

--- a/src/TestUtils/src/Microsoft.Maui.IntegrationTests/Utilities/TestEnvironment.cs
+++ b/src/TestUtils/src/Microsoft.Maui.IntegrationTests/Utilities/TestEnvironment.cs
@@ -69,11 +69,11 @@ namespace Microsoft.Maui.IntegrationTests
 			var rootDir = Environment.GetEnvironmentVariable("AGENT_TEMPDIRECTORY");
 			if (Directory.Exists(rootDir))
 			{
-				_testOutputDirectory = Path.Combine(rootDir, $"test-intg");
+				_testOutputDirectory = Path.Combine(rootDir, $"test-dir");
 			}
 			else
 			{
-				_testOutputDirectory = Path.Combine(GetMauiDirectory(), "bin", "test-intg");
+				_testOutputDirectory = Path.Combine(GetMauiDirectory(), "bin", "test-dir");
 			}
 
 			Directory.CreateDirectory(_testOutputDirectory);

